docs/ocis/adr/0015-events.md

```diff
@@ -0,0 +1,89 @@
+
+###  Example: EMail Notification
+
+A simple example is the notification feature for oCIS: Users should receive an email when another user shares a file with them. The information, that the file was shared should go out as an event from the share manager, carrying the information which file was shared to which receiver. A potential notification service that really sends out the email listens to these kind of events and sends the email out once on every received event of that specific type.
+
+## Decision Drivers
+
+- Events are supposed to simplify things and raise flexibility, also considering extensions that are not directly controlled by the ownCloud project.
+- Events should bring flexibility in the implementation of sending and receiving services.
+- Events should not obsolete other mechanisms to communicate, ie. grpc calls.
+- Sending an event has to be as little resource consuming for the sender as possible.
+- Events are never user visible.
+
+## Considered Options
+
+1. Lightweight Events with Event Queue and "At most once" QoS
+2. As 1., but with "At least once" QoS
+
+## Options
+
+### 1. Lightweight Events with Event Queue and "At most once" QoS
+
+Reva will get a messaging service that is available to all services within oCIS and Reva. It is considered as one of the mandatory services of the oCIS system. If the messaging backend is not running, neither Reva nor oCIS can be considered healthy and should shut down.
+
+All oCIS- and Reva-services can connect to the messaging bus and send so called events. The sender gets an immediate return if handing the event to the message bus was succesful or not.
+
+The sender can not make any assumptions about when the message is delivered to any receiving service, nor if zero, one or many services are listening to that event.
+
+#### Event Data
+
+Events are identified by their namespace and their respective name. The namespace is delimited by dots and starts with either "reva" or "ocis" or an future extension name. It is followed by the name of the sending service and an unique name of the event.
+
+Example: `ocis.ocdav.delete` - an event with that name sent out if an WebDAV DELETE request arrived in the oCDav service.
+
+An event can carry a payload which is encoded as json object. (See for example [NATS](https://docs.nats.io/using-nats/developer/sending/structure) ). There are no pre defined members in that object, it is fully up to the sender which data will be included in the payload. Receivers must be robust to deal with changes.
+
+#### Quality of Service
+
+Events are sent with "At most once" quality of service. That means, if a receiver is not present at the moment of publishing it might not receive the event. That requires that the receiver must have functionality to back up the situation that events were missed. Given that the event queue can be considered the backbone of the system, it is unlikely that it is not running.
```

Comment:
```suggestion
Events are sent with "At most once" quality of service. That means, if a receiver is not present at the moment of publishing it might not receive the event. That requires that the sender must have functionality to back up the situation that events were missed. Given that the event queue can be considered the backbone of the system, it is unlikely that it is not running.
```
The saga pattern puts this burden on the sender: it has to keep track of the state, eg. using a state `pending` or more to our use case `file_received`. If nobody picks up the file it has to retry later using a timeout, or we have a houskeeping process that tries to resubmit stuck steps periodically.

Reply:
Hm, no. Go back to the example with notification email: it is the notification services' concern to check if all existing shares where announced by an email. Not the shares providers.

The saga example does only apply in the case that the state of the share would be depending on the question if the notification was sent or not.

Reply:
@dragotin I think i am with both of you, but the commit suggestion from @butonic is valid.

With the "at-least-once" pattern, the sender needs to have a state machine. It needs to assume that some events are not delivered. So that introduces more state for the sender. In the saga pattern, all services need to implement a FIS pattern https://en.wikipedia.org/wiki/Finite-state_machine

Reply:
So i would commit the suggestion and add more info to the saga pattern.